src/themes/material/components/edit-chip-list/edit-chip-list.component.ts

```diff
@@ -0,0 +1,198 @@
+
+	private shortcut = new Shortcut(new ShortcutManager());
+
+	private selectedItem: ISelected = {
+		action: Actions.inactive
+	};
+
+	private incrementSelectedIndex() {
+		if (this.selectedItem.index === this.chipValues.length) {
+			this.selectedItem.index = 0;
+		} else {
+			this.selectedItem.index++;
+		}
+	}
+	private decrementSelectedIndex() {
+		if (this.selectedItem.index === 0) {
+			this.selectedItem.index = this.chipValues.length;
+		} else {
+			this.selectedItem.index--;
+		}
+	}
+
+	private indexOutOfRange(index?: number) {
+		return index === undefined || index < 0 || index > this.items.length - 1;
+	}
+
+	private deleteItemCommand = new Command({
+		canExecute: () => this.selectedItem.action === Actions.delete,
+		execute: () => {
+			this.chipValues.splice(this.selectedItem.index, 1);
+			this.decrementSelectedIndex();
+			this.initItems();
+			this.resetCurrentAction();
+			this.cdRef.detectChanges();
+		},
+	});
+
+	private addItemCommand = new Command({
+		canExecute: (value) => value,
+		execute: (value) => {
+			this.chipValues.push(value);
+			this.selectedItem.index = this.chipValues.length;
+			this.initItems();
+			this.cdRef.detectChanges();
+		},
+	});
+
+	private saveItemValueCommand = new Command({
+		canExecute: (value, index) => value && index < this.chipValues.length,
+		execute: (value, index) => {
+			this.chipValues[index] = value;
+			this.selectedItem.index = index;
+			this.initItems();
+			this.cdRef.detectChanges();
+		},
+	});
+
+	private focusPreviousCommand = new Command({
+		canExecute: () => this.selectedItem.action === Actions.focusPrevious &&
+			!this.indexOutOfRange(this.selectedItem.index),
+		execute: () => {
+			this.decrementSelectedIndex();
+			this.initItems();
+		},
+	});
+
+	private focusNextCommand = new Command({
+		canExecute: () => this.selectedItem.action === Actions.focusNext &&
+			!this.indexOutOfRange(this.selectedItem.index),
+		execute: () => {
+			this.incrementSelectedIndex();
+			this.initItems();
+		},
+	});
+
+	constructor(private cdRef: ChangeDetectorRef) {
+	}
+
+	private chipState(index: number) {
+		return index === this.chipValues.length ? ChipState.new : '';
+	}
+
+	private focusToPreviousItem(currentIndex: number) {
+		this.selectedItem = {
+			index: currentIndex,
+			action: Actions.focusPrevious
+		};
+
+		if (this.focusPreviousCommand.canExecute()) {
+			this.focusPreviousCommand.execute();
+		}
+	}
+
+	private focusToNextItem(currentIndex: number) {
+		this.selectedItem = {
+			index: currentIndex,
+			action: Actions.focusNext
+		};
+		if (this.focusNextCommand.canExecute()) {
+			this.focusNextCommand.execute();
+		}
+	}
+
+	private resetCurrentAction() {
+		this.selectedItem.action = Actions.inactive;
+	}
+
+	private saveItem(value: string | any, index: number) {
+		if (this.saveItemValueCommand.canExecute(value, index)) {
+			this.saveItemValueCommand.execute(value, index);
+		}
+	}
+
+	private deleteChip(itemIndex: number) {
+		this.selectedItem = {
```

Comment:
try to get rid of all this wrappers for the commands: this.deleteItem.execute(index) looks clear

Reply:
Do I understand you correctly: I needn't use here the pattern like:
```
                if (this.addItem.canExecute(value)) {
			this.addItem.execute(value);
		}
```
This is an atomic block of code that means a single action: addItem, so it looks like make sense to wrap it into method: addItem(). Also, most of this wrapper methods are to be embedded into the component template. It would be rather inconvenient to embed into template all this peace of code above

Reply:
So another option would be if I don't use 'canExecute()' method at all and check all conditions needed in the 'execute() method instead. Then I wouldn't need a wrapper and can just call addItem.execute() method and painlessly embed this call into the template. So, what do you think about all this?

Reply:
the action in ui should be disabled if canExecute === false, so please use it in template instead of ts code. Imagine that u are creating API for other people, it will be confusing why we have delete.execute and deleteChip together, wont it?

Reply:
Refactored this in the commit: [Merge from Dev: conflicts resolving](https://github.com/qgrid/ng2/pull/39/commits/da05f15a228a611a5059e3049c9f18402446fa8f).